SimPEG/regularization/coupling.py

```diff
@@ -0,0 +1,619 @@
+        raise NotImplementedError(
+            "The method _deriv2 has not been implemented for {}".format(
+                self.__class__.__name__
+            )
+        )
+
+    def __call__(self):
+        ''' Returns the computed value of the coupling term. '''
+        raise NotImplementedError(
+            "The method __call__ has not been implemented for {}".format(
+                self.__class__.__name__
+            )
+        )
+
+
+###############################################################################
+#                                                                             #
+#                                Cross-Gradient                               #
+#                                                                             #
+###############################################################################
+
+
+class CrossGradient(BaseCoupling):
+
+    grad_tol = 1e-10 # tolerance for avoiding the exteremly small gradient amplitude
+
+    '''
+    The cross-gradient constraint for joint inversions.
+
+    ..math::
+        \phi_c(\mathbf{m_1},\mathbf{m_2}) = \lambda \sum_{i=1}^{M} \|
+        \nabla \mathbf{m_1}_i \times \nabla \mathbf{m_2}_i \|^2
+
+    All methods assume that we are working with two models only.
+
+    '''
+    def __init__(self, mesh, indActive, mapping, **kwargs):
+
+        self.as_super.__init__(mesh, indActive, mapping, **kwargs)
+        self.map1, self.map2 = mapping.maps # Assume a map has been passed for each model.
+
+        regmesh = self.regmesh
+        self._Dx = regmesh.aveFx2CC @ regmesh.cellDiffx
+        self._Dy = regmesh.aveFy2CC @ regmesh.cellDiffy
+        if regmesh.mesh.dim == 3:
+            self._Dz = regmesh.aveFz2CC @ regmesh.cellDiffz
+
+        self._dim = regmesh.mesh.dim
+        assert self._dim in (2,3), 'Cross-Gradient is only defined for 2D or 3D'
+
+
+    def models(self, ind_models):
+        '''
+        Method to pass models to CrossGradient object and ensures models are
+        compatible for further use. Checks that models are of same size.
+
+        :param container of numpy.ndarray ind_models: [model1, model2,...]
+
+        rtype: list of numpy.ndarray models: [model1, model2,...]
+        return: models
+
+        '''
+        models = []
+        n = len(ind_models)
+        for i in range(n):
+            # check that the models are either a list, tuple, or np.ndarray
+            assert isinstance(ind_models[i], (list,tuple,np.ndarray))
+            if isinstance(ind_models[i], (list,tuple)):
+                ind_models[i] = np.array(ind_models[i])
+
+        # check if models are of same size
+        it = iter(ind_models)
+        length = len(next(it))
+        if not all(len(l)==length for l in it):
+            raise ValueError('not all models are of the same size!')
+
+        for i in range(n):
+            models.append(ind_models[i])
+
+        return models
+
+
+    def calculate_gradient(self, model, normalize=False):
```

Comment:
Below, you have the keyword `normalized` instead of `normalize`. 
I suggest being consistent and keep the same keywords.

Reply:
done